File: src/lib/api/discord.ts
import { invokeBackend } from "./errors";
import { isTauriEnv } from "./env";

export type DiscordPresenceKind = "video" | "music" | "live";

// Mirrors the backend PresencePayload (serde camelCase).
export interface DiscordPresencePayload {
  kind: DiscordPresenceKind;
  title?: string;
  subtitle?: string;
  album?: string;
  artworkUrl?: string;
  elapsedSeconds?: number;
  durationSeconds?: number;
  isPaused?: boolean;
  url?: string;
  buttonLabel?: string;
}

export async function setDiscordPresence(payload: DiscordPresencePayload): Promise<void> {
  if (!(await isTauriEnv())) return;
  try {
    await invokeBackend<void>("set_discord_presence", { payload });
  } catch (error) {
    console.warn("Failed to set Discord presence", error);
  }
}

export async function clearDiscordPresence(): Promise<void> {
  if (!(await isTauriEnv())) return;
  try {
    await invokeBackend<void>("clear_discord_presence", {});
  } catch (error) {
    console.warn("Failed to clear Discord presence", error);
  }
}

File: src-tauri/src/commands/discord.rs
use tauri::State;

use crate::errors::ErrorResponse;
use crate::services::discord_presence::{DiscordPresence, PresencePayload};

/// Push a now-playing snapshot to Discord.
#[tauri::command]
pub async fn set_discord_presence(
    payload: PresencePayload,
    presence: State<'_, DiscordPresence>,
) -> Result<(), ErrorResponse> {
    presence.set(payload);
    Ok(())
}

/// Clear the Discord activity.
#[tauri::command]
pub async fn clear_discord_presence(
    presence: State<'_, DiscordPresence>,
) -> Result<(), ErrorResponse> {
    presence.clear();
    Ok(())
}
